services/src/main/java/org/keycloak/broker/provider/IdpUserSessionExceptionHandler.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2025 Red Hat, Inc. and/or its affiliates
+ *  and other contributors as indicated by the @author tags.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package org.keycloak.broker.provider;
+
+import java.net.URI;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.Response;
+import org.jboss.logging.Logger;
+import org.keycloak.models.KeycloakContext;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ModelDuplicateException;
+import org.keycloak.models.UserSessionProvider;
+import org.keycloak.saml.BaseSAML2BindingBuilder;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class IdpUserSessionExceptionHandler implements UserSessionProvider.UserSessionExceptionHandler {
+
+    private final KeycloakContext ctx;
+
+    protected static final Logger logger = Logger.getLogger(IdpUserSessionExceptionHandler.class);
+
+    public IdpUserSessionExceptionHandler(KeycloakSession session) {
+        this.ctx = session.getContext();
+    }
+
+    @Override
+    public Response handleExceptionAtTransactionCommit(Throwable t) {
```

Comment:
Would it work if we just return an info page with a message about the user already authenticated and an option to go back to the application?

The reasoning is to avoid retrying requests (and possible issues doing so in case the re-auth fails at the IdP) as well as make the hander simpler to cover this specific corner case.